GVFS/GVFS.Service.UI/GVFSToastRequestHandler.cs

```diff
@@ -0,0 +1,137 @@
+        private const string GvfsUpgradeActionPrefix = "gvfs upgrade --confirm";
+
+        private ITracer tracer;
+        private IToastNotifier toastNotifier;
+
+        public GVFSToastRequestHandler(IToastNotifier toastNotifier, ITracer tracer)
+        {
+            this.toastNotifier = toastNotifier;
+            this.toastNotifier.UserResponseCallback = this.UserResponseCallback;
+            this.tracer = tracer;
+        }
+
+        public void HandleToastRequest(ITracer tracer, NamedPipeMessages.Notification.Request request)
+        {
+            string title = null;
+            string message = null;
+            string buttonTitle = null;
+            string args = null;
+
+            switch (request.Id)
+            {
+                case NamedPipeMessages.Notification.Request.Identifier.AutomountStart:
+                    {
+                        title = GVFSAutomountStartTitle;
+                        message = string.Format(GVFSAutomountStartMessageFormat, request.EnlistmentCount);
+                        break;
+                    }
+
+                case NamedPipeMessages.Notification.Request.Identifier.MountSuccess:
+                    {
+                        title = GVFSAutomountSuccessTitle;
+                        message = string.Format(GVFSAutomountSuccessMessageFormat, Environment.NewLine, request.Enlistment);
+                        break;
+                    }
+
+                case NamedPipeMessages.Notification.Request.Identifier.MountFailure:
+                    {
+                        title = GVFSAutomountErrorTitle;
+                        message = string.Format(GVFSAutomountErrorMessageFormat, Environment.NewLine, request.Enlistment);
+                        buttonTitle = GVFSAutomountButtonTitle;
+                        args = $"{GvfsRemountActionPrefix} {request.Enlistment}";
+                        break;
+                    }
+
+                case NamedPipeMessages.Notification.Request.Identifier.UpgradeAvailable:
+                    {
+                        title = string.Format(GVFSUpgradeTitleFormat, request.NewVersion);
+                        message = string.Format(GVFSUpgradeMessage);
+                        buttonTitle = "Upgrade";
+                        args = $"{GvfsUpgradeActionPrefix}";
+                        break;
+                    }
+            }
+
+            if (title != null && message != null)
+            {
+                this.toastNotifier.Notify(title, message, buttonTitle, args);
+            }
+        }
+
+        public void UserResponseCallback(string args)
+        {
+            if (!string.IsNullOrEmpty(args))
+            {
+                string gvfsArgs = null;
+                bool elevate = false;
+
+                if (GVFSPlatform.Instance.IsElevated())
+                {
+                    this.tracer.RelatedInfo($"GVFS.Serice.UI is currently running elevated.");

```

Comment:
Does this mean anyone can send a toast notification to `GVFS.Service.UI.exe` and have their code run elevated? I know we only accept some predefined 'actions' for the list of supported notification types, but if there are ways of breaking out, we should be careful here!

This is why I ask about escaping the 'enlistment path' args that we get back from a notification callback, if we're just shelling out to a possibly elevated cmd.exe.

Reply:
>>Does this mean anyone can send a toast notification to GVFS.Service.UI.exe and have their code run elevated?

- `GVFS.Service.UI` is running un-elevated. So it cannot. (However it runs `gvfs upgrade --confirm` command elevated after user approval).
- Also added path validation.